controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -1035,6 +1044,199 @@ public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenChec
         return result;
     }
 
+    /**
+     * The method sends a WireCommand to iterate over table keys.
+     * @param scope Stream scope.
+     * @param stream Stream name.
+     * @param suggestedKeyCount Suggested number of {@link TableKey}s to be returned by the SegmentStore.
+     * @param state Last known state of the iterator.
+     * @param hostControllerStore Host Controller store.
+     * @param clientCF Client connection factory.
+     * @param delegationToken The token to be presented to the segmentstore.
+     * @param clientRequestId Request id.
+     * @return An {@link AsyncIterator} that will return the next set of {@link TableKey}s returned from the SegmentStore.
+     */
+    public AsyncIterator<TableSegment.IteratorItem<TableKey<byte[]>>> readTableKeys(final String scope,
+                                                                                    final String stream,
+                                                                                    final int suggestedKeyCount,
+                                                                                    final IteratorState state,
+                                                                                    final HostControllerStore hostControllerStore,
+                                                                                    final ConnectionFactory clientCF,
+                                                                                    final String delegationToken,
+                                                                                    final long clientRequestId) {
+
+        final Controller.NodeUri uri = getSegmentUri(scope, stream, 0L, hostControllerStore);
+        final String qualifiedName = getScopedStreamName(scope, stream);
+        final WireCommandType type = WireCommandType.READ_TABLE_KEYS;
+        final long requestId = (clientRequestId == RequestTag.NON_EXISTENT_ID) ? idGenerator.get() : clientRequestId;
+
+        return new ContinuationTokenAsyncIterator<>(token -> {
```

Comment:
On second thoughts, maybe we dont need to use ContinuationTokenAsync iterator here. 
Instead, you can have a simple iterator: 
```AsyncIterator<Map.Entry<IteratorState, Collection<TableSegment.IteratorItem<TableKey<byte[]>>>>>```
 that, as part of each getNext returns the pair of token and next batch of entries. 

That way you dont need to depend on other PR and the api here becomes explicit that it is giving out a token and a batch of entries rather than an "iterator".

Reply:
the advantage is, if i want to get all keys, i can wrap it up in a continuation token Async Iterator. 
if i want to work only on a subset of keys, i can use the result as received by directly calling the api and working with a batch of results.. 
it will totally depend on the business logic that is interested in using these apis.. and there could be different ways it could be invoked in different places.

Reply:
Thanks @shiveshr , will open a separate PR with the updated api.